**Rationale** This assignment will help you practice running and interpreting basic regressions analysis.

**[Datasets](https://drive.google.com/drive/folders/1LJ38ctVX62ggGGygJDLlLG2V6jZ4BNml?usp=sharing)**
1. [Safegraph Temple Visitor Sample](https://drive.google.com/file/d/1u_sDxFp9fja-_cop4yzIAJ_hX2-fplWN/view?usp=sharing) 
1. [Avocado Prices](https://drive.google.com/file/d/1QJ_dyyBFkHhC6VkzVvdAFa385mtqtntM/view?usp=sharing)

**Data description** This dataset includes SafeGraph's sampled foot traffic volume for Temple University between 1-1-2019 and 5-31-2020. Safegraph collects the foot traffic data from a network of phone apps. You can view the `visits` data as the number of SafeGraph tracked devices to visit Temple University. This dataset is merged with Philadelphia's weather data. 


# 1. Run and interpret a simple linear regression (5 points)...
...to figure out the relationship between daily average temperateure as measured in degrees fahrenheit, `"temp"`, and people on campus `"visits"` for the Fall 2019 semester (limit data to M-F 8-26 - 11-24-2019, inclusive).

In your answer: 
1. First, look at data. What is the average number of SafeGraph visitors to Temple University during the teaching days of the Fall 2019 semester? Roughly what percent of the Temple student population is this?
1. Specify (write out the regression equation), run, and interpret a regression that documents the relationship between temperature and foot traffic (visits) on campus.
    1. Do more students visit campus (and presumably go to class) when the weather is warmer or colder?
    1. Besides students not wanting to go to class when the temperature is not to their liking, what else might cause the observed relationship between visits and temperature?

In [1]:
# imports here
import os, matplotlib.pyplot as plt
import pandas
#from google.colab import drive
from statsmodels.formula import api as smf
#drive.mount('drive')

In [2]:
fpath = !pwd
print(fpath)

['/home/johndoe/assign6']


In [3]:
# list directory containing the data
#fpath = 'drive/MyDrive/Teaching/Analytics/Datasets/A6/' # change this to your data folder
fpath='/home/johndoe/assign6'
os.listdir(fpath)

['temple_traffic.csv',
 'A6_F2021_Regression_1.ipynb',
 '.ipynb_checkpoints',
 'avocado.csv']

In [4]:
# read in the dataset
temple = pandas.read_csv('temple_traffic.csv', index_col = 0)
temple.head(5)

safegraph_place_id        date  visits       stn     wban  \
0  sg:00579e722c8e48178ef0c66a7c91f92c  2020-03-02      65  997286.0  99999.0   
1  sg:00579e722c8e48178ef0c66a7c91f92c  2020-03-03      68  997286.0  99999.0   
2  sg:00579e722c8e48178ef0c66a7c91f92c  2020-03-04      71  997286.0  99999.0   
3  sg:00579e722c8e48178ef0c66a7c91f92c  2020-03-05      70  997286.0  99999.0   
4  sg:00579e722c8e48178ef0c66a7c91f92c  2020-03-06      57  997286.0  99999.0   

   temp  count_temp  dewp  count_dewp     slp   ...    mxspd  gust  maxtemp  \
0  45.8        24.0   NaN         0.0  1016.8   ...      NaN   NaN     63.3   
1  53.0        24.0   NaN         0.0  1006.3   ...      NaN   NaN     59.9   
2  51.4        24.0   NaN         0.0  1006.3   ...      NaN   NaN     58.6   
3  49.2        24.0   NaN         0.0  1017.0   ...      NaN   NaN     55.8   
4  41.7        24.0   NaN         0.0  1013.1   ...      NaN   NaN     49.1   

   flag_max  mintemp  flag_min  prcp  flag_prcp  sndp  frshtt  
0         *     34.7         *   0.0          I   NaN     0.0  
1         *     45.9         *   0.0          I   NaN     0.0  
2         *     45.1         *   0.0          I   NaN     0.0  
3         *     41.7         *   0.0          I   NaN     0.0  
4         *     37.2         *   0.0          I   NaN     0.0  

[5 rows x 27 columns]

In [5]:
# convert the date column to pandas datetime using pd.to_datetime(...)
# if you've fogotten, look through notes from module 3
temple['date'] = pandas.to_datetime(temple['date'])
temple.head(5)

safegraph_place_id       date  visits       stn     wban  \
0  sg:00579e722c8e48178ef0c66a7c91f92c 2020-03-02      65  997286.0  99999.0   
1  sg:00579e722c8e48178ef0c66a7c91f92c 2020-03-03      68  997286.0  99999.0   
2  sg:00579e722c8e48178ef0c66a7c91f92c 2020-03-04      71  997286.0  99999.0   
3  sg:00579e722c8e48178ef0c66a7c91f92c 2020-03-05      70  997286.0  99999.0   
4  sg:00579e722c8e48178ef0c66a7c91f92c 2020-03-06      57  997286.0  99999.0   

   temp  count_temp  dewp  count_dewp     slp   ...    mxspd  gust  maxtemp  \
0  45.8        24.0   NaN         0.0  1016.8   ...      NaN   NaN     63.3   
1  53.0        24.0   NaN         0.0  1006.3   ...      NaN   NaN     59.9   
2  51.4        24.0   NaN         0.0  1006.3   ...      NaN   NaN     58.6   
3  49.2        24.0   NaN         0.0  1017.0   ...      NaN   NaN     55.8   
4  41.7        24.0   NaN         0.0  1013.1   ...      NaN   NaN     49.1   

   flag_max  mintemp  flag_min  prcp  flag_prcp  sndp  frshtt  
0         *     34.7         *   0.0          I   NaN     0.0  
1         *     45.9         *   0.0          I   NaN     0.0  
2         *     45.1         *   0.0          I   NaN     0.0  
3         *     41.7         *   0.0          I   NaN     0.0  
4         *     37.2         *   0.0          I   NaN     0.0  

[5 rows x 27 columns]

In [6]:
# create a new column called 'dow' (for day of week) using temple.date.dt.weekday
# note that Monday through Sunday corresponds to 0 through 6
temple['dow'] = temple.date.dt.weekday
temple.head(5)


safegraph_place_id       date  visits       stn     wban  \
0  sg:00579e722c8e48178ef0c66a7c91f92c 2020-03-02      65  997286.0  99999.0   
1  sg:00579e722c8e48178ef0c66a7c91f92c 2020-03-03      68  997286.0  99999.0   
2  sg:00579e722c8e48178ef0c66a7c91f92c 2020-03-04      71  997286.0  99999.0   
3  sg:00579e722c8e48178ef0c66a7c91f92c 2020-03-05      70  997286.0  99999.0   
4  sg:00579e722c8e48178ef0c66a7c91f92c 2020-03-06      57  997286.0  99999.0   

   temp  count_temp  dewp  count_dewp     slp ...   gust  maxtemp  flag_max  \
0  45.8        24.0   NaN         0.0  1016.8 ...    NaN     63.3         *   
1  53.0        24.0   NaN         0.0  1006.3 ...    NaN     59.9         *   
2  51.4        24.0   NaN         0.0  1006.3 ...    NaN     58.6         *   
3  49.2        24.0   NaN         0.0  1017.0 ...    NaN     55.8         *   
4  41.7        24.0   NaN         0.0  1013.1 ...    NaN     49.1         *   

   mintemp  flag_min  prcp  flag_prcp  sndp  frshtt  dow  
0     34.7         *   0.0          I   NaN     0.0    0  
1     45.9         *   0.0          I   NaN     0.0    1  
2     45.1         *   0.0          I   NaN     0.0    2  
3     41.7         *   0.0          I   NaN     0.0    3  
4     37.2         *   0.0          I   NaN     0.0    4  

[5 rows x 28 columns]

In [7]:
# select only the rows where the date lies between 2019/8/26 and 2019/11/24 (inclusive) 
# and correspond to a regular school date (e.g. are M-F)
# call the reuslting dataframe F2019
# remember you can select rows using df.loc[conditions]
# If you forgot how to do this, look at the notes and in class exercise from module 3

F2019 = temple.sort_values(by = 'date')
F2019 = F2019[F2019['date'] >= '2019/8/26'] 
F2019 = F2019[F2019['date'] <= '2019/11/24']
F2019 = F2019.loc[(F2019.dow < 5)] 
F2019


safegraph_place_id       date  visits       stn  \
140  sg:00579e722c8e48178ef0c66a7c91f92c 2019-08-26     166  997286.0   
141  sg:00579e722c8e48178ef0c66a7c91f92c 2019-08-27     162  997286.0   
142  sg:00579e722c8e48178ef0c66a7c91f92c 2019-08-28     172  997286.0   
143  sg:00579e722c8e48178ef0c66a7c91f92c 2019-08-29     168  997286.0   
144  sg:00579e722c8e48178ef0c66a7c91f92c 2019-08-30     194  997286.0   
175  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-02      88  997286.0   
176  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-03     155  997286.0   
177  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-04     183  997286.0   
178  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-05     172  997286.0   
179  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-06     130  997286.0   
112  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-09     177  997286.0   
113  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-10     146  997286.0   
114  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-11     173  997286.0   
115  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-12     173  997286.0   
116  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-13     187  997286.0   
49   sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-16     163  997286.0   
50   sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-17     161  997286.0   
51   sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-18     162  997286.0   
52   sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-19     164  997286.0   
53   sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-20     153  997286.0   
7    sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-23     169  997286.0   
8    sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-24     182  997286.0   
9    sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-25     156  997286.0   
10   sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-26     168  997286.0   
11   sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-27     165  997286.0   
231  sg:00579e722c8e48178ef0c66a7c91f92c 2019-09-30     136  997286.0   
232  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-01     123  997286.0   
233  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-02     127  997286.0   
234  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-03     126  997286.0   
235  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-04     128  997286.0   
..                                   ...        ...     ...       ...   
434  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-14     121  997286.0   
435  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-15     132  997286.0   
436  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-16     131  997286.0   
437  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-17     134  997286.0   
438  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-18     137  997286.0   
420  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-21     168  997286.0   
421  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-22     147  997286.0   
422  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-23     134  997286.0   
423  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-24     146  997286.0   
424  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-25     126  997286.0   
294  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-28     143  997286.0   
295  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-29     110  997286.0   
296  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-30     107  997286.0   
297  sg:00579e722c8e48178ef0c66a7c91f92c 2019-10-31     116  997286.0   
298  sg:00579e722c8e48178ef0c66a7c91f92c 2019-11-01     134  997286.0   
56   sg:00579e722c8e48178ef0c66a7c91f92c 2019-11-04     113  997286.0   
57   sg:00579e722c8e48178ef0c66a7c91f92c 2019-11-05     105  997286.0   
58   sg:00579e722c8e48178ef0c66a7c91f92c 2019-11-06     109  997286.0   
59   sg:00579e722c8e48178ef0c66a7c91f92c 2019-11-07     104  997286.0   
60   sg:00579e722c8e48178ef0c66a7c91f92c 2019-11-08     111  997286.0   
504  sg:00579e722c8e48178ef0c66a7c91f92c 2019-11-11     117  997286.0   
505  sg:00579e722c8e48178ef0c66a7c91f92c 2019-11-12     106  997286.0   
506  sg:00579e722c8e48178ef0c66a7c91f92c 2019-11-13     106  997286.0   
507  sg

In [8]:
# find the maximum date of F2019, make sure that it is less than or equal to 2019/11/24
max_value = F2019['date'].max()



In [9]:
# find the minimum date of F2019, make sure that it is greater than or equal to 2019/8/26
min_value = F2019['date'].min()
min_value


Timestamp('2019-08-26 00:00:00')

In [10]:
# display the unique values of dow in F2019 (use the .unique() method on dow column)
# make sure there are only M-F (e.g. 0-4)
F2019['dow'].unique()


array([0, 1, 2, 3, 4])

In [11]:
# compute the mean visits
mean = F2019['visits'].mean()
mean

137.73846153846154

In [12]:
# look up the total number of temple students in 2019 and compute the percentage the above average represents
temple_students = 40000
percentage = (mean/temple_students)  * 100
percentage



0.34434615384615386

**Edit this cell**
The average number of visitors in the sample is **137.738**, this is approximately **0.34%** of the Temple student population.

**Edit this cell**

Before running a regression to document the relationship between temperature and the sampled visitors, answer the following.

The logical dependent variable should be **vistors**, the logical indepdent (aka explanatory) variable should be **temperature** because **temperatrue** should cause changes in **visitors** and not the other way around.


Consequently, the following regression equation describes the relationship between temperature and foot traffic (change the $x$ and $y$ to the approriate variable names):

$$
y = \beta_0 +\beta_1x + e 
$$


In [24]:
# run the above regression using statsmodels
# call the resulting regression results object the variable "res"

res = smf.ols('visits ~ temp', data = F2019).fit()


In [25]:
# print the regression summary table (e.g. print(res.summary()))
##print(res.summary())
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 visits   R-squared:                       0.468
Model:                            OLS   Adj. R-squared:                  0.459
Method:                 Least Squares   F-statistic:                     55.33
Date:                Tue, 19 Oct 2021   Prob (F-statistic):           3.43e-10
Time:                        21:22:05   Log-Likelihood:                -285.36
No. Observations:                  65   AIC:                             574.7
Df Residuals:                      63   BIC:                             579.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     40.8564     13.255      3.082      0.003      14.369      67.343
temp           1.5446      0.208      7.438      0.000       1.130       1.960
==============================================================================
Omnibus:                        7.362   Durbin-Watson:                   1.341
Prob(Omnibus):                  0.025   Jarque-Bera (JB):                6.733
Skew:                          -0.639   Prob(JB):                       0.0345
Kurtosis:                       3.922   Cond. No.                         344.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Edit this cell**

The regression's R$^2$ is **0.468**, this suggests that **47%**\% of the variation in **visits** is explained by **temperature**.

The intercept ($\beta_0$) is **40.85**, this means that when **temperature** is 0 **(degrees, fahrenheit)**, **visits** is on average **41 people**. This is statistically **significant**, meaning that it is different from **the average visits** in the population.

The coefficient ($\beta_1$) for ** visits ** is **.468 **. This means that for every ** 1 degree ** increase in ** temp **, **visits ** **increases (choose one)** by **.468**. This is statistically **significant**, meaning that it is different from **random visits** in the population. In other words, we can reject the null hypothesis that there is no effect of temperature on campus foot traffic.  **Additionally, because the P> |t| is .003, this means that there is a less than 1 percent chance that the temp variable has no effect on the dependent variable, visits.** 

# 2. Avocado price and demand (5 points) (multiple regression)

For this exercise, use our familiar avocado dataset.

For the following cities, run a multiple regression that investigates the relationship between demand and price while simultaneously accounting for the effect of regions.

Use only the conventional avocado data from the following cities:

```
cities = ['Philadelphia', 'NewYork', 'BaltimoreWashington', 'Boston', 'Chicago', 'SanFrancisco']
```


In [16]:
# read in the dataset, call it avocado
avocado = pandas.read_csv('avocado.csv', index_col = 0)
avocado

Date  AveragePrice  Total Volume       4046       4225     4770  \
0      2015-12-27          1.33      64236.62    1036.74   54454.85    48.16   
1      2015-12-20          1.35      54876.98     674.28   44638.81    58.33   
2      2015-12-13          0.93     118220.22     794.70  109149.67   130.50   
3      2015-12-06          1.08      78992.15    1132.00   71976.41    72.58   
4      2015-11-29          1.28      51039.60     941.48   43838.39    75.78   
5      2015-11-22          1.26      55979.78    1184.27   48067.99    43.61   
6      2015-11-15          0.99      83453.76    1368.92   73672.72    93.26   
7      2015-11-08          0.98     109428.33     703.75  101815.36    80.00   
8      2015-11-01          1.02      99811.42    1022.15   87315.57    85.34   
9      2015-10-25          1.07      74338.76     842.40   64757.44   113.00   
10     2015-10-18          1.12      84843.44     924.86   75595.85   117.07   
11     2015-10-11          1.28      64489.17    1582.03   52677.92   105.32   
12     2015-10-04          1.31      61007.10    2268.32   49880.67   101.36   
13     2015-09-27          0.99     106803.39    1204.88   99409.21   154.84   
14     2015-09-20          1.33      69759.01    1028.03   59313.12   150.50   
15     2015-09-13          1.28      76111.27     985.73   65696.86   142.00   
16     2015-09-06          1.11      99172.96     879.45   90062.62   240.79   
17     2015-08-30          1.07     105693.84     689.01   94362.67   335.43   
18     2015-08-23          1.34      79992.09     733.16   67933.79   444.78   
19     2015-08-16          1.33      80043.78     539.65   68666.01   394.90   
20     2015-08-09          1.12     111140.93     584.63  100961.46   368.95   
21     2015-08-02          1.45      75133.10     509.94   62035.06   741.08   
22     2015-07-26          1.11     106757.10     648.75   91949.05   966.61   
23     2015-07-19          1.26      96617.00    1042.10   82049.40  2238.02   
24     2015-07-12          1.05     124055.31     672.25   94693.52  4257.64   
25     2015-07-05          1.35     109252.12     869.45   72600.55  5883.16   
26     2015-06-28          1.37      89534.81     664.23   57545.79  4662.71   
27     2015-06-21          1.27     104849.39     804.01   76688.55  5481.18   
28     2015-06-14          1.32      89631.30     850.58   55400.94  4377.19   
29     2015-06-07          1.07     122743.06     656.71   99220.82    90.32   
...           ...           ...           ...        ...        ...      ...   
18219  2018-02-11          1.56    1317000.47   98465.26  270798.27  1839.80   
18220  2018-02-04          1.53    1384683.41  117922.52  287724.61  1703.52   
18221  2018-01-28          1.61    1336979.09  118616.17  280080.34  1270.61   
18222  2018-01-21          1.63    1283987.65  108705.28  259172.13  1490.02   
18223  2018-01-14          1.59    1476651.08  145680.62  323669.83  1580.01   
18224  2018-01-07          1.51    1517332.70  129541.43  296490.29  1289.07   
18225  2018-03-25          1.60     271723.08   26996.28   77861.39   117.56   
18226  2018-03-18          1.73     210067.47   33437.98   47165.54   110.40   
18227  2018-03-11          1.63     264691.87   27566.25   60383.57   276.42   
18228  2018-03-04          1.46     347373.17   25990.60   71213.19    79.01   
18229  2018-02-25          1.49     301985.61   34200.18   49139.34    85.58   
18230  2018-02-18          1.64     224798.60   30149.00   38800.64   123.13   
18231  2018-02-11          1.47     275248.53   24732.55   61713.53   243.00   
18232  2018-02-04          1.41     283378.47   22474.66   55360.49   133.41   
18233  2018-01-28          1.80     185974.53   22918.40   33051.14    93.52   
18234  2018-01-21          1.83     189317.99   27049.44   33561.32   439.47   
18235  2018-01-14          1.82     207999.67   33869.12   47435.14   433.52   
18236  2018-01-07          1.48     297190.60   34734.97   62967.74   157.77   
18237  2018-0

In [17]:
# select only the rows corresponding to the cities listed above
# call the resulting dataframe avocado (replace avocado with the result of this subset selection)
# use .isin(...) method (see notes from module 4 bar plot to find an example of this)
# also make sure to select only the data for the conventional avocado type
cities = ['Philadelphia', 'NewYork', 'BaltimoreWashington', 'Boston', 'Chicago', 'SanFrancisco']
avocado = avocado.loc[avocado['region'].isin(cities)] 
avocado = avocado.loc[avocado['type'] == 'conventional']
avocado


Date  AveragePrice  Total Volume       4046        4225      4770  \
104   2015-12-27          1.17     596819.40   40450.49   394104.02  17353.79   
105   2015-12-20          1.11     642682.40   44741.37   434590.82  19112.14   
106   2015-12-13          1.15     619509.33   44400.26   399851.59  24291.20   
107   2015-12-06          1.04     649141.25   51129.29   453586.50  25760.49   
108   2015-11-29          1.16     545800.05   41028.15   379156.09  20147.53   
109   2015-11-22          1.12     572019.80   40867.80   386137.78  23343.49   
110   2015-11-15          1.04     718330.48   47109.75   503470.08  37990.29   
111   2015-11-08          1.05     830967.23   58579.27   576512.71  51035.04   
112   2015-11-01          1.12     742220.67   51860.07   473819.14  58615.76   
113   2015-10-25          1.19     656892.03   53766.25   397911.35  49085.74   
114   2015-10-18          0.99    1005603.78   59121.55   745385.26  49171.33   
115   2015-10-11          1.24     653560.14   34707.02   392008.60  58490.10   
116   2015-10-04          1.28     657444.04   35342.34   407609.95  42257.73   
117   2015-09-27          1.19     696281.81   34309.05   451061.03  59018.61   
118   2015-09-20          1.17     709268.62   39619.87   450283.01  45336.12   
119   2015-09-13          1.09     924993.61   55699.25   675524.33  42270.47   
120   2015-09-06          1.18     823623.36   63903.73   522568.44  68180.35   
121   2015-08-30          1.24     734721.29   48455.25   458207.46  42754.34   
122   2015-08-23          1.16     783935.36   59242.02   499100.39  47216.18   
123   2015-08-16          1.14     882754.72   86492.03   597184.73  43102.61   
124   2015-08-09          1.16     821523.45   56362.75   557363.94  42933.87   
125   2015-08-02          1.17     800603.23   53957.63   535167.10  44749.58   
126   2015-07-26          1.20     772926.12   52746.41   496508.65  45010.81   
127   2015-07-19          1.18     800194.34   52954.94   496206.47  50319.84   
128   2015-07-12          1.10     850931.58   58827.99   547537.41  50679.01   
129   2015-07-05          1.16     896298.77   66504.67   567640.70  56044.51   
130   2015-06-28          1.15     862261.64   65051.95   534535.21  51109.85   
131   2015-06-21          1.16     862840.73   58398.76   527866.50  51952.74   
132   2015-06-14          1.09     955046.82   64005.15   593499.63  48344.86   
133   2015-06-07          1.26     794310.29   62585.45   462951.05  43720.95   
...          ...           ...           ...        ...         ...       ...   
8820  2018-02-11          1.27    2051389.99  169218.18  1319012.42   1222.29   
8821  2018-02-04          1.28    2959541.38   96084.63  2250967.25   3253.28   
8822  2018-01-28          1.20    2278728.69   82652.82  1501055.47    745.50   
8823  2018-01-21          1.27    2135242.76   46444.43  1642533.16   1165.10   
8824  2018-01-14          1.67    1294149.71   39681.84   818051.50   1338.06   
8825  2018-01-07          1.67    1287480.31   39884.07   866953.05    573.69   
8862  2018-03-25          1.31     503277.65   43421.10   252219.75    248.04   
8863  2018-03-18          1.33     471175.79   44696.37   241775.09    334.21   
8864  2018-03-11          1.32     486492.24   45869.80   237689.35    297.86   
8865  2018-03-04          1.36     443651.36   37703.03   231605.34    335.17   
8866  2018-02-25          1.26     490292.38   55979.65   235447.72    314.32   
8867  2018-02-18          1.24     443312.68   67616.46   233345.32    236.22   
8868  2018-02-11          1.02     611269.34  130162.76   320845.90    213.80   
8869  2018-02-04          1.12     819224.30   62044.22   467927.80    645.87   
8870  2018-01-28          1.23     537883.88   90381.50   256525.12    390.07   
8871  2018-01-21          1.32     527860.09   69862.88   303263.33    383.99   
8872  2018-01-14          1.50     419189.36   36351.30   232190.19    443.69   
8873  2018-01-07          1.56     408203.4

In [18]:
# show the unique regions in the resulting dataframe 
# make sure we have only the ones we wanted to select from above

avocado['region'].unique()results



array(['BaltimoreWashington', 'Boston', 'Chicago', 'NewYork',
       'Philadelphia', 'SanFrancisco'], dtype=object)

In [19]:
# show the unique types in the resulting dataframe 
# make sure we have only conventional

avocado['type'].unique()


array(['conventional'], dtype=object)

**edit this cell**
To model the demand of avocaldos, we want to run the following regression:

$$
Demand = \beta_0 + \beta_1 Price +\beta_2 Boston +\beta_3Chicago +\beta_4NewYork + \beta_5 Philadelphia + \beta_6 SanFrancisco +e
$$

In this equation (given the cities we've selected for the dataset), $\beta_0$ represents the average demand in **in each city** when price is \$ $0$.



In [20]:
# since statsmodels expects variable names to not have spaces
# use a list comprehension to replace  spaces with underscores
# make sure you understand how/why this works:
avocado.columns = [c.replace(' ', '') for c in avocado.columns]
avocado

Date  AveragePrice  TotalVolume       4046        4225      4770  \
104   2015-12-27          1.17    596819.40   40450.49   394104.02  17353.79   
105   2015-12-20          1.11    642682.40   44741.37   434590.82  19112.14   
106   2015-12-13          1.15    619509.33   44400.26   399851.59  24291.20   
107   2015-12-06          1.04    649141.25   51129.29   453586.50  25760.49   
108   2015-11-29          1.16    545800.05   41028.15   379156.09  20147.53   
109   2015-11-22          1.12    572019.80   40867.80   386137.78  23343.49   
110   2015-11-15          1.04    718330.48   47109.75   503470.08  37990.29   
111   2015-11-08          1.05    830967.23   58579.27   576512.71  51035.04   
112   2015-11-01          1.12    742220.67   51860.07   473819.14  58615.76   
113   2015-10-25          1.19    656892.03   53766.25   397911.35  49085.74   
114   2015-10-18          0.99   1005603.78   59121.55   745385.26  49171.33   
115   2015-10-11          1.24    653560.14   34707.02   392008.60  58490.10   
116   2015-10-04          1.28    657444.04   35342.34   407609.95  42257.73   
117   2015-09-27          1.19    696281.81   34309.05   451061.03  59018.61   
118   2015-09-20          1.17    709268.62   39619.87   450283.01  45336.12   
119   2015-09-13          1.09    924993.61   55699.25   675524.33  42270.47   
120   2015-09-06          1.18    823623.36   63903.73   522568.44  68180.35   
121   2015-08-30          1.24    734721.29   48455.25   458207.46  42754.34   
122   2015-08-23          1.16    783935.36   59242.02   499100.39  47216.18   
123   2015-08-16          1.14    882754.72   86492.03   597184.73  43102.61   
124   2015-08-09          1.16    821523.45   56362.75   557363.94  42933.87   
125   2015-08-02          1.17    800603.23   53957.63   535167.10  44749.58   
126   2015-07-26          1.20    772926.12   52746.41   496508.65  45010.81   
127   2015-07-19          1.18    800194.34   52954.94   496206.47  50319.84   
128   2015-07-12          1.10    850931.58   58827.99   547537.41  50679.01   
129   2015-07-05          1.16    896298.77   66504.67   567640.70  56044.51   
130   2015-06-28          1.15    862261.64   65051.95   534535.21  51109.85   
131   2015-06-21          1.16    862840.73   58398.76   527866.50  51952.74   
132   2015-06-14          1.09    955046.82   64005.15   593499.63  48344.86   
133   2015-06-07          1.26    794310.29   62585.45   462951.05  43720.95   
...          ...           ...          ...        ...         ...       ...   
8820  2018-02-11          1.27   2051389.99  169218.18  1319012.42   1222.29   
8821  2018-02-04          1.28   2959541.38   96084.63  2250967.25   3253.28   
8822  2018-01-28          1.20   2278728.69   82652.82  1501055.47    745.50   
8823  2018-01-21          1.27   2135242.76   46444.43  1642533.16   1165.10   
8824  2018-01-14          1.67   1294149.71   39681.84   818051.50   1338.06   
8825  2018-01-07          1.67   1287480.31   39884.07   866953.05    573.69   
8862  2018-03-25          1.31    503277.65   43421.10   252219.75    248.04   
8863  2018-03-18          1.33    471175.79   44696.37   241775.09    334.21   
8864  2018-03-11          1.32    486492.24   45869.80   237689.35    297.86   
8865  2018-03-04          1.36    443651.36   37703.03   231605.34    335.17   
8866  2018-02-25          1.26    490292.38   55979.65   235447.72    314.32   
8867  2018-02-18          1.24    443312.68   67616.46   233345.32    236.22   
8868  2018-02-11          1.02    611269.34  130162.76   320845.90    213.80   
8869  2018-02-04          1.12    819224.30   62044.22   467927.80    645.87   
8870  2018-01-28          1.23    537883.88   90381.50   256525.12    390.07   
8871  2018-01-21          1.32    527860.09   69862.88   303263.33    383.99   
8872  2018-01-14          1.50    419189.36   36351.30   232190.19    443.69   
8873  2018-01-07          1.56    408203.49   22384.92   239185.79    407.29   
8982  2018-0

In [21]:
# take a look at the resulting dataframe

avocado.head()

Date  AveragePrice  TotalVolume      4046       4225      4770  \
104  2015-12-27          1.17    596819.40  40450.49  394104.02  17353.79   
105  2015-12-20          1.11    642682.40  44741.37  434590.82  19112.14   
106  2015-12-13          1.15    619509.33  44400.26  399851.59  24291.20   
107  2015-12-06          1.04    649141.25  51129.29  453586.50  25760.49   
108  2015-11-29          1.16    545800.05  41028.15  379156.09  20147.53   

     TotalBags  SmallBags  LargeBags  XLargeBags          type  year  \
104  144911.10  142543.88    2367.22         0.0  conventional  2015   
105  144238.07  141751.91    2486.16         0.0  conventional  2015   
106  150966.28  149070.40    1895.88         0.0  conventional  2015   
107  118664.97  117112.77    1552.20         0.0  conventional  2015   
108  105468.28  103615.95    1852.33         0.0  conventional  2015   

                  region  
104  BaltimoreWashington  
105  BaltimoreWashington  
106  BaltimoreWashington  
107  BaltimoreWashington  
108  BaltimoreWashington

In [114]:
##regions = {'b_region': 'Boston', 'bw_region': 'BaltimoreWashington', 'c_region': 'Chicago', 'n_region': 'NewYork', 'p_region': 'Philadelphia', 's_region': 'SanFrancisco'}
##for key in regions:
##    avocado[key] = 0  
##    avocado.loc[avocado['region'] == regions[key], key] = 1
##avocado
#regions = ['b_region': 'Boston', 'bw_region': 'BaltimoreWashington', 'c_region': 'Chicago', 'n_region': 'NewYork', 'p_region': 'Philadelphia', 's_region': 'SanFrancisco'}
##for key in regions:
##    avocado[key] = 0  
##    avocado.loc[avocado['region'] == regions[key], key] = avocado['AveragePrice']
##avocado


Date  AveragePrice  TotalVolume       4046        4225      4770  \
104   2015-12-27          1.17    596819.40   40450.49   394104.02  17353.79   
105   2015-12-20          1.11    642682.40   44741.37   434590.82  19112.14   
106   2015-12-13          1.15    619509.33   44400.26   399851.59  24291.20   
107   2015-12-06          1.04    649141.25   51129.29   453586.50  25760.49   
108   2015-11-29          1.16    545800.05   41028.15   379156.09  20147.53   
109   2015-11-22          1.12    572019.80   40867.80   386137.78  23343.49   
110   2015-11-15          1.04    718330.48   47109.75   503470.08  37990.29   
111   2015-11-08          1.05    830967.23   58579.27   576512.71  51035.04   
112   2015-11-01          1.12    742220.67   51860.07   473819.14  58615.76   
113   2015-10-25          1.19    656892.03   53766.25   397911.35  49085.74   
114   2015-10-18          0.99   1005603.78   59121.55   745385.26  49171.33   
115   2015-10-11          1.24    653560.14   34707.02   392008.60  58490.10   
116   2015-10-04          1.28    657444.04   35342.34   407609.95  42257.73   
117   2015-09-27          1.19    696281.81   34309.05   451061.03  59018.61   
118   2015-09-20          1.17    709268.62   39619.87   450283.01  45336.12   
119   2015-09-13          1.09    924993.61   55699.25   675524.33  42270.47   
120   2015-09-06          1.18    823623.36   63903.73   522568.44  68180.35   
121   2015-08-30          1.24    734721.29   48455.25   458207.46  42754.34   
122   2015-08-23          1.16    783935.36   59242.02   499100.39  47216.18   
123   2015-08-16          1.14    882754.72   86492.03   597184.73  43102.61   
124   2015-08-09          1.16    821523.45   56362.75   557363.94  42933.87   
125   2015-08-02          1.17    800603.23   53957.63   535167.10  44749.58   
126   2015-07-26          1.20    772926.12   52746.41   496508.65  45010.81   
127   2015-07-19          1.18    800194.34   52954.94   496206.47  50319.84   
128   2015-07-12          1.10    850931.58   58827.99   547537.41  50679.01   
129   2015-07-05          1.16    896298.77   66504.67   567640.70  56044.51   
130   2015-06-28          1.15    862261.64   65051.95   534535.21  51109.85   
131   2015-06-21          1.16    862840.73   58398.76   527866.50  51952.74   
132   2015-06-14          1.09    955046.82   64005.15   593499.63  48344.86   
133   2015-06-07          1.26    794310.29   62585.45   462951.05  43720.95   
...          ...           ...          ...        ...         ...       ...   
8820  2018-02-11          1.27   2051389.99  169218.18  1319012.42   1222.29   
8821  2018-02-04          1.28   2959541.38   96084.63  2250967.25   3253.28   
8822  2018-01-28          1.20   2278728.69   82652.82  1501055.47    745.50   
8823  2018-01-21          1.27   2135242.76   46444.43  1642533.16   1165.10   
8824  2018-01-14          1.67   1294149.71   39681.84   818051.50   1338.06   
8825  2018-01-07          1.67   1287480.31   39884.07   866953.05    573.69   
8862  2018-03-25          1.31    503277.65   43421.10   252219.75    248.04   
8863  2018-03-18          1.33    471175.79   44696.37   241775.09    334.21   
8864  2018-03-11          1.32    486492.24   45869.80   237689.35    297.86   
8865  2018-03-04          1.36    443651.36   37703.03   231605.34    335.17   
8866  2018-02-25          1.26    490292.38   55979.65   235447.72    314.32   
8867  2018-02-18          1.24    443312.68   67616.46   233345.32    236.22   
8868  2018-02-11          1.02    611269.34  130162.76   320845.90    213.80   
8869  2018-02-04          1.12    819224.30   62044.22   467927.80    645.87   
8870  2018-01-28          1.23    537883.88   90381.50   256525.12    390.07   
8871  2018-01-21          1.32    527860.09   69862.88   303263.33    383.99   
8872  2018-01-14          1.50    419189.36   36351.30   232190.19    443.69   
8873  2018-01-07          1.56    408203.49   22384.92   239185.79    407.29   
8982  2018-0

In [26]:
# run the regression using statsmodels
# call the result object res
# for demand, use 'Total Bags'

res = smf.ols('TotalBags ~ AveragePrice + region', data = avocado).fit()

In [40]:
# print the regression summary table
res.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              TotalBags   R-squared:                       0.765
Model:                            OLS   Adj. R-squared:                  0.764
Method:                 Least Squares   F-statistic:                     546.7
Date:                Tue, 19 Oct 2021   Prob (F-statistic):          1.30e-312
Time:                        22:02:21   Log-Likelihood:                -12529.
No. Observations:                1014   AIC:                         2.507e+04
Df Residuals:                    1007   BIC:                         2.511e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept               1.847e+05   1.05e+04     17.671      0.000    1.64e+05    2.05e+05
region[T.Boston]       -8.097e+04   6144.983    -13.177      0.000    -9.3e+04   -6.89e+04
region[T.Chicago]      -9.941e+04   6141.100    -16.188      0.000   -1.11e+05   -8.74e+04
region[T.NewYork]       1.843e+05   6152.184     29.961      0.000    1.72e+05    1.96e+05
region[T.Philadelphia]  -6.29e+04   6149.588    -10.229      0.000    -7.5e+04   -5.08e+04
region[T.SanFrancisco] -1.097e+05   6149.942    -17.839      0.000   -1.22e+05   -9.76e+04
AveragePrice            8121.8105   7073.444      1.148      0.251   -5758.568     2.2e+04
==============================================================================
Omnibus:                      586.949   Durbin-Watson:                   0.765
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            10971.908
Skew:                           2.262   Prob(JB):                         0.00
Kurtosis:                      18.467   Cond. No.                         12.9
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [78]:
model = smf.ols('TotalBags ~ AveragePrice + region', data = avocado).fit()

In [81]:
params = model.params
params['AveragePrice']

-405297.78276384296

**Edit this cell**
1. The R$^2$ is **.765**, this means that **77%**\% of the variation in demand is explained by the model.
1. The intercept ($\beta_0$) is **_184,700** and is statistically **significant **. This is the average demand in **_Total Bags_** when price is \$0.
1. The coefficient for Boston is **_essentially zero_**, which means that the average demand in Boston is **_____slightly _** **fewer** bags than in **__on average_**. This difference is statistically **insignificant (pick one)**
1. The coefficient for price is **_8121.81_**, which means that when price increases by \$1 in any region, demand increases by on average **____8122_______** bags. This effect is statistically **significant** which means that there is enough evidence to show that price **does ** have a positive** impact on demand.
1. The region with the highest demand is **_New York_**.

# 3. Putting what we've learned all together (the power of programming) (2 points - extra point for bonus, try your best - there may or may not be a prize involved)

In this exercise, you will run one regression for each region and each avocado type, i.e. one regression of total bags sold on price for Philly conventional, one for Philly organic, one for NY conventional, one for NY organic, and so on... (think loop). 

You will then visualize the coefficients for conventional and organic price by region on a scatter plot (each dot's (x, y) coordinates = one region's (price coefficient for conventional, price coefficient for organic). 

To do this break down the problem to the following components (and then put it all together):

* Suppose you have selected a region $r$ and avocado type $t$, how would you run the regression?
* How can you extract just the price coefficient given a regression result object `res`?
* For some region, `r`, and some avocado type `t`, how would you select only the rows pertaining to region `r` and type `t`?
* How can you loop through regions and types? (Need 2 nested "for" loops, one inside the other)
* Suppose you have named the conventional and organic price coefficients for region `r` as `bConv` and `bOrg`, how would you create a list that looks like: `[region, conventional coeff, organic coefficient]`?
* Given a list `results`, how would you *append* the list above to `results`? i.e. make keep adding to `results`: `[[region1, conventional coeff1, organic coeff1], [region2, conventional coeff2, organic coeff2], [region3, conventional coeff3, organic coeff3],...]`
* Once you've run every regression and have a complete `results` list, convert it to a dataframe using `df = pd.DataFrame(results)`, give it some column names.

The final code might have a structure similar to the pseudo code below:

```
results = list()
for r in ##all the unique regions##:
    for t in ##all the unique types##:
        temp = # Select the appropriate rows of df
        # run the regression using temp
        if ##this iteration represents conventional avocados##:
            ##extract the price parameter assign it to the variable bConv##
        else:
            ##extract the price parameter and name it bOrg##
    ##append the results with newest [region, conventional price coeff, organic price coefficient]##
df = pd.DataFrame(results) # convert to dataframe
df.columns = ['region', 'beta_conv', 'beta_org'] # rename dataframe columns
```

The resulting dataframe of coefficients should look something like:

| region | beta_conv | beta_org |
|-|-|-|
| Albany |8.579340e+02	 |-3168.491384|
| Atlanta| -3.377365e+05 | -6352.937420 |
| BaltimoreWashington | -3.251570e+05 | -15537.003211 |
|...|...|...|...|

(note numbers above are illustrative, not necessarily representative of actual results)


In [28]:
# first re-read the original avocado dataframe to use for this problem
avocado = pandas.read_csv('avocado.csv', index_col = 0)

In [30]:
# rename columns to get rid of spaces
avocado.columns = [c.replace(' ', '') for c in avocado.columns]
avocado.head(2)

Date  AveragePrice  TotalVolume     4046      4225   4770  TotalBags  \
0  2015-12-27          1.33     64236.62  1036.74  54454.85  48.16    8696.87   
1  2015-12-20          1.35     54876.98   674.28  44638.81  58.33    9505.56   

   SmallBags  LargeBags  XLargeBags          type  year  region  
0    8603.62      93.25         0.0  conventional  2015  Albany  
1    9408.07      97.49         0.0  conventional  2015  Albany

In [34]:
unique_regions = avocado['region'].unique()
unique_regions
unique_types = avocado['type'].unique()
unique_types

array(['conventional', 'organic'], dtype=object)

In [83]:
# run all the regressions
# create a dataframe with region, beta_conv, beta_org as the 3 columns 
# using the regression results
results = []

for a_region in unique_regions: 
    beta_conv = 'na'
    beta_org = 'na'
    for kind in unique_types:
        temp_df = avocado.loc[(avocado.region == a_region)]
        temp_df = temp_df.loc[(temp_df.type == kind)]
        model = smf.ols('TotalBags ~ AveragePrice + region', data = temp_df).fit()
        params = model.params
        ##beta_key = 'region[T.' + a_region + ']'
        average_price = params['AveragePrice']
        ##print(beta)
        if kind == 'conventional':
            ##beta_conv = model.rsquared
            beta_conv = average_price
        if kind == 'organic':
            ##beta_org = model.rsquared
            beta_org = average_price
    temp_list = [a_region, beta_conv, beta_org]
    results.append(temp_list)


    
df = pandas.DataFrame(results) # convert to dataframe
df.columns = ['region', 'beta_conv', 'beta_org'] # rename dataframe columns
df

region     beta_conv       beta_org
0                Albany  8.579340e+02   -3168.491384
1               Atlanta -8.373534e+04   -5041.977163
2   BaltimoreWashington -1.472740e+03  -14715.844065
3                 Boise -2.862924e+04   -1438.258267
4                Boston  7.050460e+03    -369.330085
5      BuffaloRochester -7.171370e+04   -6817.766138
6            California -8.462063e+05  -19414.846809
7             Charlotte  2.788339e+03   -1339.038227
8               Chicago -2.209711e+04    2780.475253
9      CincinnatiDayton -3.001482e+04  -12285.334723
10             Columbus -2.150627e+04   -7673.401216
11        DallasFtWorth  3.282912e+04  -14727.711026
12               Denver -2.951864e+05  -33328.763388
13              Detroit -4.761408e+04  -16349.719575
14          GrandRapids -1.570927e+04   -5307.265152
15           GreatLakes  1.188229e+05  -86671.747091
16   HarrisburgScranton -5.763337e+03  -13967.292160
17  HartfordSpringfield  5.071228e+03   -5225.013443
18              Houston -9.930901e+04  -18528.090267
19         Indianapolis -1.971945e+04   -4856.981965
20         Jacksonville -3.565409e+04   -1498.597456
21             LasVegas -5.135322e+04   -3598.797323
22           LosAngeles -7.048588e+05  -27498.705762
23           Louisville -5.693065e+03   -2604.180215
24    MiamiFtLauderdale -7.735051e+04   -2128.434417
25             Midsouth  2.171263e+05   21814.337090
26            Nashville -5.716375e+04   -5925.443070
27     NewOrleansMobile -3.712780e+04   -4638.384459
28              NewYork  9.602090e+04  -24363.709569
29            Northeast  1.525947e+05 -129186.372092
30   NorthernNewEngland  9.086072e+02   -7432.154713
31              Orlando -6.214428e+04   -2887.644846
32         Philadelphia -7.193851e+03  -14877.989666
33        PhoenixTucson  5.694786e+04   -3057.136947
34           Pittsburgh -2.449085e+04  -10660.052566
35               Plains  1.299340e+04  -22371.411504
36             Portland -3.559129e+05  -16087.769797
37    RaleighGreensboro  6.865955e+02     727.274656
38      RichmondNorfolk  2.287446e+04   -2715.476217
39              Roanoke -1.392618e+04   -3864.696048
40           Sacramento -1.745360e+04    -339.349660
41             SanDiego -1.115811e+05   -1675.515127
42         SanFrancisco -1.643547e+03    1004.879672
43              Seattle -2.986643e+05  -24973.992871
44        SouthCarolina  1.277277e+04   -6906.597116
45         SouthCentral  8.733649e+04  -71593.694275
46            Southeast -4.290777e+05  -27656.530608
47              Spokane -4.223723e+04   -1041.106019
48              StLouis -1.100739e+05   -4424.537144
49             Syracuse -1.318647e+04   -3426.537166
50                Tampa -9.564444e+04   -2091.179250
51              TotalUS -3.202584e+05  246283.396284
52                 West -1.063483e+06  -99223.974231
53     WestTexNewMexico -6.617689e+03   -8610.230656

In [ ]:
# Get rid of the TotalUS region from the regression results dataframe



In [ ]:
# show the full resulting table



In [ ]:
# make scatter plot of the coefficients
# change aesthetics so the dots are not a single blob




In [ ]:
# which regions have upward sloping demand curve for organic (as price increases, so did demand)?
# use code to show these rows of df



In [ ]:
# which regions have upward sloping demand curve for conventional (as price increases, so did demand)?
# use code to show these rows of df



Why might the relationship between price and demand be upward sloping instead of the expected downward sloping demand curve theorized in economics? i.e. the law of demand, when price increases, demand decreases. 

**Your answer here**